Allow python protobuf codegen export to coexist with namespace/host package

achrafmam2 opened the pull request:
Before this change, if a codegen export clashes with an existing package in the venv the export halts completely. The current change skips the problematic exports. This is sub optimal because not all symbols are exported but better than halting the export.

See #21659

tgolsson commented:
Nice! Unfortunately I think the logging here will not be visible, since it would be coming from a child process launched by Pants if I read this code directly -- and thus eaten and not displayed in the common case. I wonder if using `shutil.copytree(...., dirs_exist_ok=True)` is the simpler solution that solves this permanently?

achrafmam2 commented:
Good idea @tgolsson. Done.

tgolsson commented:
Perfect! Before I unblock CI, I'll have to ask you to add some release notes as well here... otherwise CI will yell at you anyways :) 

https://github.com/pantsbuild/pants/blob/main/docs/notes/2.25.x.md

achrafmam2 commented:
@tgolsson thanks for walking me through it. I updated the release notes.

Also note that I updated the code, because when I tested locally it failed. There is extra step that cleans up the `codegen_dir` which is unfortunately not tested. Maybe once I have more time I can add more coverage.

tdyas commented:
>  I forced myself though and it failed. Maybe it's worth to force new authors to do it.

Thanks for testing! (Even small changes can be larger than they seem ...)

tgolsson commented:
I've been erring on whether we should backport this to 2.24.x and even possibly 2.23.x, what do you think?

tgolsson commented:
@achrafmam2 We have a bot that does the PRs etc. Asked in #development on slack, I'm not sure how we do the notes now. The notes update from *this branch* also shouldn't be backported.
